File: src/copilot_console/app/models/message.py
"""Message models."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class MessageStep(BaseModel):
    """A step in the assistant's response (tool call, reasoning, etc.)."""

    title: str
    detail: str | None = None


class MessageAttachment(BaseModel):
    """Attachment on a message (from SDK history)."""

    type: str = "file"
    path: str | None = None
    displayName: str | None = None


class Message(BaseModel):
    """Chat message."""

    id: str
    # Durable SDK anchor when available (e.g., for pins/reasoning joins).
    sdk_message_id: str | None = None
    role: Literal["user", "assistant"]
    content: str
    timestamp: datetime
    steps: list[MessageStep] | None = None
    attachments: list[MessageAttachment] | None = None


class AttachmentRef(BaseModel):
    """Reference to an uploaded file for SDK attachment."""

    type: Literal["file", "directory"] = "file"
    path: str
    displayName: str | None = None


class MessageCreate(BaseModel):
    """Request to send a message."""

    content: str = Field(..., min_length=0)
    is_new_session: bool = False  # True if this is the first message in a new session
    mode: Literal["enqueue", "immediate"] | None = None  # Message delivery mode
    attachments: list[AttachmentRef] | None = None  # File/directory attachments
    agent_mode: str | None = None  # Agent mode to set before sending (interactive/plan/autopilot)
    fleet: bool = False  # True to use fleet mode (parallel sub-agents)
    compact: bool = False  # True to compact context before sending (deferred from new/resumed session)
    agent: str | None = None  # Agent name to select before sending (deferred from new/resumed session)


class MessageDelta(BaseModel):
    """Streaming message delta."""

    content: str


class MessageComplete(BaseModel):
    """Message completion event."""

    message_id: str

File: frontend/src/types/message.ts
export interface ChatStep {
  title: string;
  detail?: string;
}

export interface MessageAttachment {
  type: string;
  path?: string;
  displayName?: string;
}

export interface Message {
  id: string;
  /** Durable SDK anchor when available (used for pins/permalinks). */
  sdk_message_id?: string;
  /** SDK event UUID for truncate/fork operations. */
  event_id?: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: string;
  steps?: ChatStep[];
  mode?: 'enqueue' | 'immediate';
  attachments?: MessageAttachment[];
  /**
   * Side-channel marker. Messages with `kind === 'help'` are rendered with a
   * distinct visual identity (amber ❓ avatar, "Help Agent" label, badge) to
   * make clear they are NOT part of the primary agent's conversation context.
   */
  kind?: 'help';
}

export interface StreamingMessage {
  id: string;
  role: 'assistant';
  content: string;
  isStreaming: boolean;
  steps?: ChatStep[];
}
